feat: Add text border support and fix background rotation bug

EedgeY opened the pull request:
# feat: Add text border and padding properties + fix background rotation bug

## Overview
Implements border and padding properties for text elements as requested in issue #851, and fixes a background rotation positioning bug.

## Key Features
- Individual border width control for each side (top, right, bottom, left)
- Border color picker for unified border styling
- Individual padding control for each side
- Seamless integration with existing text properties
- Consistent UI/PDF rendering
- Fixed background rotation positioning bug

## Changes

### Border and Padding Implementation
- Extended `TextSchema` type with `borderWidth`, `borderColor`, and `padding` properties
- Added `renderBorder` function in PDF rendering with proper coordinate calculations
- Updated UI rendering with CSS border and padding styles
- Enhanced property panel with new border and padding controls
- Maintains compatibility with existing text features



https://github.com/user-attachments/assets/b6785a18-9b1f-4476-986d-642dbe6f5de8




### Bug Fixes
- Fixed background rotation positioning where rotated backgrounds were not properly aligned
- Improved coordinate calculation for rotated elements in PDF rendering



https://github.com/user-attachments/assets/77bd3c13-009b-49e7-a29c-0b7be5d0548c




Closes #851

hand-dot commented:
@EedgeY

Please pull main again and run `npm run test:ui:update-snapshots` in the root directory to update the snapshots, which should make the CI pass.

hand-dot reviewed (changes requested):
@EedgeY 
Please fix the following two points:
* Overall modifications are needed from a code reusability perspective
* Need to pass CI

EedgeY commented:
# Eliminate Code Duplication and Unify Border Functionality

## Overview

This PR makes the pdfme codebase more compact by eliminating code duplication between text and table schemas and unifying border rendering functionality.

## Key Changes

### 1. Added Common Utility Functions (`packages/schemas/src/utils.ts`)

#### New Type Definition
```typescript
export interface BoxDimensions {
  top: number;
  right: number;
  bottom: number;
  left: number;
}
```

#### Common Border Rendering Function
```typescript
export const renderBorder = (
  page: PDFPage,
  x: number,
  y: number,
  width: number,
  height: number,
  borderWidth: BoxDimensions,
  borderColor: string,
  rotate: Rotation,
  opacity: number,
  pageHeight: number,
  schema: { position: { x: number; y: number } },
  colorType?: ColorType,
) => { /* ... */ }
```

#### Common PropPanel Function
```typescript
export const getBoxDimensionProp = (step = 1, i18n?: (key: string) => string) => {
  // Generates box dimension properties for borderWidth, padding, etc.
}
```

#### Common UI Function
```typescript
export const createBorderLineDiv = (
  width: string,
  height: string,
  top: string | null,
  right: string | null,
  bottom: string | null,
  left: string | null,
  borderColor: string,
) => { /* ... */ }
```

### 2. Text Schema Modifications

#### `packages/schemas/src/text/pdfRender.ts`
- **Removed**: Duplicated `renderBorder` function (68 lines)
- **Changed**: Now uses common `renderBorder` function from utils
- **Removed**: Unnecessary imports (`PDFPage`, `Rotation`, `Spacing`)

```typescript
// Before
const renderBorder = (/* 68 lines of duplicated code */) => { /* ... */ }

// After  
import { renderBorder } from '../utils.js';
renderBorder(/* ... */);
```

#### `packages/schemas/src/text/propPanel.ts`
- **Changed**: `borderWidth` and `padding` properties now use common `getBoxDimensionProp` function
- **Removed**: Duplicated box dimension property definitions (~50 lines)

```typescript
// Before
borderWidth: {
  title: i18n('schemas.borderWidth') || 'Border Width',
  type: 'object',
  widget: 'card',
  column: 4,
  properties: {
    top: { /* individual definition */ },
    right: { /* individual definition */ },
    bottom: { /* individual definition */ },
    left: { /* individual definition */ },
  },
},

// After
borderWidth: {
  title: i18n('schemas.borderWidth') || 'Border Width',
  type: 'object',
  widget: 'card',
  column: 4,
  properties: getBoxDimensionProp(0.1, i18n),
},
```

### 3. Table Schema Modifications

#### `packages/schemas/src/tables/helper.ts`
- **Removed**: Duplicated `getBoxDimensionProp` function (15 lines)
- **Changed**: Now imports and uses common `getBoxDimensionProp` function

```typescript
// Before
const getBoxDimensionProp = (step = 1) => {
  // 15 lines of duplicated code
};

// After
import { getBoxDimensionProp } from '../utils.js';
```


## Related Issues

This change addresses the code duplication mentioned in the PR feedback, specifically:
- Sharing border rendering logic with `/packages/schemas/src/tables/cell.ts`
- Simplifying `packages/schemas/src/text/propPanel.ts` by using `getBoxDimensionProp` from `packages/schemas/src/tables/helper.ts`

hand-dot reviewed (changes requested):
I understand what you want to do, but it seems you don't fully understand the intent of my comment.
packages/schemas/src/tables/cell.ts and the text renderer should be able to share much more in common.
In other words, now that the text renderer supports padding and border, the cell schema should become much simpler. In fact, the cell schema should be able to be eliminated altogether.
Also, please don't make me a middleman with AI. Reviews are taking up my time.

EedgeY commented:
This is the maximum extent of changes I can implement at this point.
Based on the previous feedback, I have made the following changes:
Changes Made
Added common utilities (utils.ts):
BoxDimensions interface for consistent border/padding definitions
renderBorder() function for unified border rendering
getBoxDimensionProp() and createBorderLineDiv() utility functions
Eliminated code duplication:
Removed duplicated renderBorder function from text schema (68 lines)
Removed duplicated getBoxDimensionProp from table helper (15 lines)
Simplified property panel definitions using shared utilities (~50 lines)
Removed cell schema entirely:
BREAKING CHANGE: Deleted packages/schemas/src/tables/cell.ts (~200+ lines)
Cell functionality now provided by unified text schema with border/padding support
CellStyle now extends BaseTextStyle for consistency
Unified type system:
Text and table schemas now share common Spacing and border/padding interfaces
Single source of truth for border rendering logic
Total Impact
~400+ lines of duplicated code eliminated
Cell schema removed as requested in previous feedback
Unified border/padding functionality across all schemas
Please review and provide feedback on this implementation.

hand-dot commented:
Hi @EedgeY 

When combining padding and border with dynamic font size, the font size setting doesn't take the padding and border into account, causing the text to overflow from the box.
![CleanShot 2025-06-16 at 12 00 01](https://github.com/user-attachments/assets/36ee2722-48c3-4c3e-a554-b772670bbb1b)

[template (2).json](https://github.com/user-attachments/files/20749340/template.2.json)

EedgeY commented:
This is an existing bug that was already there.
I've made some minor fixes, but we might need to make more fundamental changes. Let me think about this a bit more.

pdfme.com/playground
<img width="1508" alt="スクリーンショット 2025-06-16 12 21 42" src="https://github.com/user-attachments/assets/c5bba261-2450-4963-8439-4c80050954bc" />



local
<img width="1508" alt="スクリーンショット 2025-06-16 12 21 56" src="https://github.com/user-attachments/assets/31080ac5-0551-4563-b384-04e90919f65f" />

EedgeY commented:
The issue is becoming clearer.
It seems like pdfRender calculates correctly and has logic to fit within the bounds, but uiRender isn't calculating properly?
I'll work on getting a precise understanding of the problem.


https://github.com/user-attachments/assets/d688ddf3-ee19-49b3-917f-facec8a53a02
